test/errtest.c

```diff
@@ -148,6 +148,100 @@ static int raised_error(void)
     return 1;
 }
 
+static int test_marks(void)
```

Comment:
There should be test cases where errors are actually popped off of the error stack when popping back to the mark.  Likewise, there should be some testing for clearing the last mark in the presence of errors.  Without these, the code to search back through the stack is never exercised.

For the nested pop case, this means multiple errors need to be raised: at least one at each level popped.

It would be good to see cases where more than one error needs to be passed over before finding a mark.

Ideally, the tests should exercise the error stack wrapping around the array but I'm not going to insist on this since it's going to be a kludge.

Reply:
> There should be test cases where errors are actually popped off of the error stack when popping back to the mark.

There is such a test. It's this one:

````
    /* Test popping errors */
    if (!TEST_true(ERR_set_mark()))
        return 0;
    CRYPTOerr(0, ERR_R_INTERNAL_ERROR);
    if (!TEST_ulong_ne(mallocfail, ERR_peek_last_error())
            || !TEST_true(ERR_pop_to_mark())
            || !TEST_ulong_eq(mallocfail, ERR_peek_last_error()))
        return 0;
````

> Likewise, there should be some testing for clearing the last mark in the presence of errors.

Again, there is one. Its this one:

````
    /* Setting and clearing a mark should not affect the errors on the stack */
    if (!TEST_true(ERR_set_mark()))
        return 0;
    CRYPTOerr(0, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
    if (!TEST_true(ERR_clear_last_mark())
            || !TEST_ulong_eq(shouldnot, ERR_peek_last_error()))
        return 0;
````

> For the nested pop case, this means multiple errors need to be raised: at least one at each level popped.

Yep, I have this one too:

````
    if (!TEST_true(ERR_set_mark()))
        return 0;
    CRYPTOerr(0, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
    shouldnot = ERR_peek_last_error();
    if (!TEST_ulong_ne(mallocfail, shouldnot)
            || !TEST_true(ERR_set_mark()))
        return 0;
    CRYPTOerr(0, ERR_R_INTERNAL_ERROR);
    if (!TEST_ulong_ne(shouldnot, ERR_peek_last_error())
            || !TEST_true(ERR_pop_to_mark())
            || !TEST_ulong_eq(shouldnot, ERR_peek_last_error())
            || !TEST_true(ERR_pop_to_mark())
            || !TEST_ulong_eq(mallocfail, ERR_peek_last_error()))
        return 0;
````

> It would be good to see cases where more than one error needs to be passed over before finding a mark.

I don't have this one. I can add that.